src/Components/Admin/ObjectEditor/keypadpuzzle.js

```diff
@@ -0,0 +1,121 @@
+import React, { useState } from "react";
+import AddIcon from "@material-ui/icons/Add";
+import { Button, IconButton } from "@material-ui/core";
+import { DeleteForever } from "@material-ui/icons";
+import TextField from "@material-ui/core/TextField";
+import { makeStyles } from "@material-ui/core/styles";
+
+const useStyles = makeStyles(() => ({
+    input: {
+        "&:invalid": {
+            border: "red solid 2px",
+        },
+    },
+}));
+
+export default function KeypadPuzzle(props) {
+    const classes = useStyles();
+    const [tempPass, setTempPass] = useState(props.pass);
+    const [pass, setPass] = useState(props.pass);
+    const [showNumpadPrompt, setNumpadPrompt] = useState(false);
+    const [showKeyboardPrompt, setKeyboardPrompt] = useState(false);
+    const [showSavePass, setSavePass] = useState(false);
+
+    const showPrompt = () => {
+        if (props.isNumpad) {
+            setNumpadPrompt(true);
+        } else {
+            setKeyboardPrompt(true);
+        }
+        setSavePass(true);
+    };
+
+    const deletePass = () => {
+        props.savePass("");
+        setPass("");
+        if (props.isNumpad) {
+            setNumpadPrompt(false);
+        } else {
+            setKeyboardPrompt(false);
+        }
+        setSavePass(false);
+    };
+
+    const savePass = () => {
+        props.savePass(tempPass);
+        setPass(tempPass);
+    };
+
+    const numpadOnChange = (e) => {
+        const re = /^[0-9\b]{1,7}$/;
+        if (e.target.value === "" || re.test(e.target.value)) {
+            setTempPass(e.target.value);
+            setSavePass(true);
+        } else {
+            alert(
+                "Error: Password must be at most 7 characters and only consist of numbers"
+            );
```

Comment:
Todo in future: remove alerts and use dynamic html elements to show error. ie. have an error boolean, have some red text hidden when boolean is false, and show when it is true. instead of alert here, we toggle the boolean.